explainability-core/src/main/java/org/kie/trustyai/explainability/local/shap/background/CounterfactualGenerator.java

```diff
@@ -0,0 +1,220 @@
+        RealVector goalVector = MatrixUtilsExtensions.vectorFromPredictionOutput(goal);
+        List<Pair<Integer, Double>> distances = new ArrayList<>();
+        for (int i = 0; i < seedOutputs.size(); i++) {
+            distances.add(new Pair<>(i, MatrixUtilsExtensions.vectorFromPredictionOutput(
+                    new PredictionOutput(seedOutputs.get(i).getOutputs())).getDistance(goalVector)));
+        }
+        distances.sort(Comparator.comparingDouble(Pair::getValue));
+        return IntStream.range(0, n).mapToObj(i -> seeds.get(distances.get(i).getKey())).collect(Collectors.toList());
+    }
+
+    // BUILDER FOR CF GENERATOR ========================================================================================
+    public static Builder builder(List<PredictionInput> seeds, PredictionProvider model, PredictionOutput goal) {
+        return new Builder(seeds, model, goal);
+    }
+
+    public static class Builder {
+        private final List<PredictionInput> seeds;
+        private Integer kSeeds = null;
+        private final PredictionProvider model;
+        private CounterfactualConfig counterfactualConfig;
+        private final PredictionOutput goal;
+        private PerturbationContext pc = new PerturbationContext(new Random(), 0);
+        private long timeoutSeconds = 30;
+        private long stepCount = 30_000L;
+        private double goalThreshold = 0.1;
+
+        private Builder(List<PredictionInput> seeds, PredictionProvider model, PredictionOutput goal) {
+            this.seeds = seeds;
+            this.model = model;
+            this.goal = goal;
+        }
+
+        public Builder withKSeeds(int kSeeds) {
+            this.kSeeds = kSeeds;
+            return this;
+        }
+
+        public Builder withCounterfactualConfig(CounterfactualConfig counterfactualConfig) {
+            this.counterfactualConfig = counterfactualConfig;
+            return this;
+        }
+
+        public Builder withPerturbationContext(PerturbationContext pc) {
+            this.pc = pc;
+            return this;
+        }
+
+        public Builder withTimeoutSeconds(long timeoutSeconds) {
+            this.timeoutSeconds = timeoutSeconds;
+            return this;
+        }
+
+        public Builder withStepCount(long stepCount) {
+            this.stepCount = stepCount;
+            return this;
+        }
+
+        public Builder withGoalThreshold(double goalThreshold) {
+            this.goalThreshold = goalThreshold;
+            return this;
+        }
+
+        public CounterfactualGenerator build() {
+            final TerminationConfig terminationConfig =
+                    new TerminationConfig().withScoreCalculationCountLimit(this.stepCount);
+            final SolverConfig solverConfig = SolverConfigBuilder
+                    .builder().withTerminationConfig(terminationConfig).build();
+            solverConfig.setRandomSeed(0L);
```

Comment:
I would remove having a fixed seed and `REPRODUCIBLE` by default.
If this is needed (e.g. for testing), it could be passed via `withCounterfactualConfig`, perhaps.

Reply:
Okay, will do.